### PR TITLE
Integrate Cohere re-ranking API

### PR DESCRIPTION
**Pull request by rafax**
Integrates Cohere re-ranking [API](https://cohere.com/rerank) for server-side Cody Context ([RFC 969](https://linear.app/sourcegraph/project/v1-of-two-stage-intent-detection-context-retrieval-system-c4f7093e9eab/overview)). 
Before this PR, we only supported `identity` ranker (which returned all items in the input order), which is still the default choice (when Cohere API key is not provided). 

Closes https://linear.app/sourcegraph/issue/AI-134/add-non-poc-ranking

## Test plan

- tested locally, use 
```
"cody.serverSideContext": {
      "reranker": {
        "type": "cohere",
        "apiKey": "TOKEN"
      }
    }
```
to test locally

### CODE REVIEW

File: schema/site.schema.json [hunk 1]
```diff
@@ -656,6 +656,16 @@
                   }
                 }
               }
+            },
+            "reranker" : {
+              "description": "Reranker to use for rankContext requests",
+              "type": "object",
+              "properties": {
```
Comment (janhartman) [inline schema/site.schema.json:663]: Can we make this a bit more futureproof? I'd add a string corresponding to the `CodyReranker` type here and then use that one so we don't rely on behavior like what's described below.

Comment (rafax) [reply schema/site.schema.json:663]: Yeah, I'm actually changing this to allow specifying the model too, thank you for quick turnaround on PR review ;)

Comment (rafax) [reply schema/site.schema.json:663]: And, to comment on your specific concern, it'd be ideal to do something like `oneOf` in Protocol buffers - I'll see if that's possible using JSON schema

Comment (rafax) [reply schema/site.schema.json:663]: @janhartman PTAL now

**Review by janhartman — APPROVED** [review]
Nice!

**Pull request merged**